File: facebook-common/src/test/java/com/facebook/login/CustomTabLoginMethodHandlerTest.java
/*
 * Copyright (c) 2014-present, Facebook, Inc. All rights reserved.
 *
 * You are hereby granted a non-exclusive, worldwide, royalty-free license to use,
 * copy, modify, and distribute this software in source code or binary form for use
 * in connection with the web services and APIs provided by Facebook.
 *
 * As with any software that integrates with the Facebook platform, your use of
 * this software is subject to the Facebook Developer Principles and Policies
 * [http://developers.facebook.com/policy/]. This copyright notice shall be
 * included in all copies or substantial portions of the software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
 * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
 * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
 * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
 * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

package com.facebook.login;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertNull;
import static org.mockito.ArgumentMatchers.nullable;
import static org.mockito.Matchers.any;
import static org.mockito.Matchers.anyInt;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.powermock.api.mockito.PowerMockito.mockStatic;
import static org.powermock.api.mockito.PowerMockito.when;

import android.content.Context;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.content.pm.ResolveInfo;
import android.content.pm.ServiceInfo;
import android.os.Bundle;
import androidx.fragment.app.Fragment;
import com.facebook.AccessToken;
import com.facebook.AccessTokenSource;
import com.facebook.AuthenticationToken;
import com.facebook.FacebookActivity;
import com.facebook.FacebookException;
import com.facebook.FacebookOperationCanceledException;
import com.facebook.FacebookSdk;
import com.facebook.TestUtils;
import com.facebook.internal.FetchedAppSettings;
import com.facebook.internal.FetchedAppSettingsManager;
import com.facebook.internal.Utility;
import com.facebook.internal.Validate;
import com.facebook.internal.security.OidcSecurityUtil;
import java.security.PublicKey;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;
import org.junit.Before;
import org.junit.Test;
import org.mockito.ArgumentCaptor;
import org.powermock.api.mockito.PowerMockito;
import org.powermock.core.classloader.annotations.PowerMockIgnore;
import org.powermock.core.classloader.annotations.PrepareForTest;

@PowerMockIgnore({"org.mockito.*", "org.robolectric.*", "org.powermock.*"})
@PrepareForTest({
  LoginClient.class,
  Validate.class,
  Utility.class,
  FacebookSdk.class,
  AccessToken.class,
  FetchedAppSettings.class,
  FetchedAppSettingsManager.class,
  OidcSecurityUtil.class
})
public class CustomTabLoginMethodHandlerTest extends LoginHandlerTestCase {
  private static final String SIGNED_REQUEST_STR =
      "ggarbage.eyJhbGdvcml0aG0iOiJITUFDSEEyNTYiLCJ"
          + "jb2RlIjoid2h5bm90IiwiaXNzdWVkX2F0IjoxNDIyNTAyMDkyLCJ1c2VyX2lkIjoiMTIzIn0";
  private static final String CHROME_PACKAGE = "com.android.chrome";
  private static final String DEV_PACKAGE = "com.chrome.dev";
  private static final String BETA_PACKAGE = "com.chrome.beta";

  private LoginClient.Request request;

  @Before
  public void setUp() {
    mockTryAuthorize();
    request = createRequest();
    PowerMockito.mockStatic(FacebookSdk.class);
    PowerMockito.when(FacebookSdk.getApplicationId())
        .thenReturn(AuthenticationTokenTestUtil.APP_ID);

    // mock and bypass signature verification
    PowerMockito.mockStatic(OidcSecurityUtil.class);
    when(OidcSecurityUtil.getRawKeyFromEndPoint(any(String.class))).thenReturn("key");
    when(OidcSecurityUtil.getPublicKeyFromString(any(String.class)))
        .thenReturn(PowerMockito.mock(PublicKey.class));
    when(OidcSecurityUtil.verify(any(PublicKey.class), any(String.class), any(String.class)))
        .thenReturn(true);
  }

  @Test
  public void testCustomTabHandlesSuccess() {
    testCustomTabHandles(getEncodedAuthTokenString());
  }

  @Test
  public void testCustomTabHandlesSuccessWithEmptyAuthenticationToken() {
    LoginClient.Result result = testCustomTabHandles("");

    AuthenticationToken authenticationToken = result.authenticationToken;
    assertEquals(null, authenticationToken);
  }

  @Test
  public void testCustomTabHandlesSuccessWithNoAuthenticationToken() {
    LoginClient.Result result = testCustomTabHandles(null);

    AuthenticationToken authenticationToken = result.authenticationToken;
    assertEquals(null, authenticationToken);
  }

  @Test
  public void testIdTokenWithNonceCustomTabHandlesSuccess() {
    mockCustomTabRedirectActivity(true);
    LoginClient.Request requestWithNonce = createRequestWithNonce();
    CustomTabLoginMethodHandler handler = new CustomTabLoginMethodHandler(mockLoginClient);

    String expectedIdTokenString = AuthenticationTokenTestUtil.getEncodedAuthTokenStringForTest();
    final Bundle bundle = new Bundle();
    bundle.putString("access_token", ACCESS_TOKEN);
    bundle.putString(AuthenticationToken.AUTHENTICATION_TOKEN_KEY, expectedIdTokenString);
    bundle.putString("expires_in", String.format("%d", EXPIRES_IN_DELTA));
    bundle.putString("signed_request", SIGNED_REQUEST_STR);
    handler.onComplete(requestWithNonce, bundle, null);

    final ArgumentCaptor<LoginClient.Result> resultArgumentCaptor =
        ArgumentCaptor.forClass(LoginClient.Result.class);
    verify(mockLoginClient, times(1)).completeAndValidate(resultArgumentCaptor.capture());

    final LoginClient.Result result = resultArgumentCaptor.getValue();
    assertNotNull(result);
    assertEquals(LoginClient.Result.Code.SUCCESS, result.code);

    final AuthenticationToken idToken = result.authenticationToken;
    assertNotNull(idToken);
    assertEquals(expectedIdTokenString, idToken.getToken());

    final AccessToken token = result.token;
    assertNotNull(token);
    assertEquals(ACCESS_TOKEN, token.getToken());
    assertDateDiffersWithinDelta(new Date(), token.getExpires(), EXPIRES_IN_DELTA * 1000, 1000);
    TestUtils.assertSamePermissions(PERMISSIONS, token.getPermissions());
  }

  @Test
  public void testIGCustomTabHandlesSuccess() {
    mockCustomTabRedirectActivity(true);
    LoginClient.Request igRequest = createIGWebRequest();
    CustomTabLoginMethodHandler handler = new CustomTabLoginMethodHandler(mockLoginClient);

    final Bundle bundle = new Bundle();
    bundle.putString("access_token", ACCESS_TOKEN);
    bundle.putString("graph_domain", "instagram");
    bundle.putString("signed_request", SIGNED_REQUEST_STR);
    handler.onComplete(igRequest, bundle, null);

    final ArgumentCaptor<LoginClient.Result> resultArgumentCaptor =
        ArgumentCaptor.forClass(LoginClient.Result.class);
    verify(mockLoginClient, times(1)).completeAndValidate(resultArgumentCaptor.capture());

    final LoginClient.Result result = resultArgumentCaptor.getValue();
    assertNotNull(result);
    assertEquals(LoginClient.Result.Code.SUCCESS, result.code);

    final AccessToken token = result.token;
    assertNotNull(token);
    assertEquals(ACCESS_TOKEN, token.getToken());
    assertEquals(USER_ID, token.getUserId());
    assertEquals("instagram", token.getGraphDomain());
    assertEquals(AccessTokenSource.INSTAGRAM_CUSTOM_CHROME_TAB, token.getSource());
    TestUtils.assertSamePermissions(PERMISSIONS, token.getPermissions());
  }

  @Test
  public void testCustomTabHandlesCancel() {
    mockCustomTabRedirectActivity(true);
    CustomTabLoginMethodHandler handler = new CustomTabLoginMethodHandler(mockLoginClient);

    handler.onComplete(request, null, new FacebookOperationCanceledException());

    final ArgumentCaptor<LoginClient.Result> resultArgumentCaptor =
        ArgumentCaptor.forClass(LoginClient.Result.class);
    verify(mockLoginClient, times(1)).completeAndValidate(resultArgumentCaptor.capture());
    final LoginClient.Result result = resultArgumentCaptor.getValue();

    assertNotNull(result);
    assertEquals(LoginClient.Result.Code.CANCEL, result.code);
    assertNull(result.token);
    assertNotNull(result.errorMessage);
  }

  @Test
  public void testCustomTabHandlesError() {
    mockCustomTabRedirectActivity(true);
    CustomTabLoginMethodHandler handler = new CustomTabLoginMethodHandler(mockLoginClient);

    handler.onComplete(request, null, new FacebookException(ERROR_MESSAGE));

    final ArgumentCaptor<LoginClient.Result> resultArgumentCaptor =
        ArgumentCaptor.forClass(LoginClient.Result.class);
    verify(mockLoginClient, times(1)).completeAndValidate(resultArgumentCaptor.capture());
    final LoginClient.Result result = resultArgumentCaptor.getValue();

    assertNotNull(result);
    assertEquals(LoginClient.Result.Code.ERROR, result.code);
    assertNull(result.token);
    assertNotNull(result.errorMessage);
    assertEquals(ERROR_MESSAGE, result.errorMessage);
  }

  @Test
  public void testTryAuthorizeNeedsRedirectActivity() {
    mockChromeCustomTabsSupported(true, CHROME_PACKAGE);
    mockCustomTabRedirectActivity(true);

    CustomTabLoginMethodHandler handler = new CustomTabLoginMethodHandler(mockLoginClient);

    assertEquals(handler.tryAuthorize(request), 1);
  }

  @Test
  public void testTryAuthorizeWithChromePackage() {
    mockCustomTabRedirectActivity(true);
    mockChromeCustomTabsSupported(true, CHROME_PACKAGE);

    CustomTabLoginMethodHandler handler = new CustomTabLoginMethodHandler(mockLoginClient);
    assertEquals(handler.tryAuthorize(request), 1);
  }

  @Test
  public void testTryAuthorizeWithChromeBetaPackage() {
    mockCustomTabRedirectActivity(true);
    mockChromeCustomTabsSupported(true, BETA_PACKAGE);

    CustomTabLoginMethodHandler handler = new CustomTabLoginMethodHandler(mockLoginClient);
    assertEquals(handler.tryAuthorize(request), 1);
  }

  @Test
  public void testTryAuthorizeWithChromeDevPackage() {
    mockCustomTabRedirectActivity(true);
    mockChromeCustomTabsSupported(true, DEV_PACKAGE);

    CustomTabLoginMethodHandler handler = new CustomTabLoginMethodHandler(mockLoginClient);

    assertEquals(handler.tryAuthorize(request), 1);
  }

  private void mockTryAuthorize() {
    mockStatic(FacebookSdk.class);
    when(FacebookSdk.isInitialized()).thenReturn(true);
    mockStatic(AccessToken.class);
    when(AccessToken.getCurrentAccessToken()).thenReturn(null);
    Fragment fragment = mock(LoginFragment.class);
    when(mockLoginClient.getFragment()).thenReturn(fragment);
  }

  private void mockChromeCustomTabsSupported(final boolean supported, final String packageName) {
    final List<ResolveInfo> resolveInfos = new ArrayList<>();
    ResolveInfo resolveInfo = new ResolveInfo();
    ServiceInfo serviceInfo = new ServiceInfo();
    serviceInfo.packageName = packageName;
    resolveInfo.serviceInfo = serviceInfo;
    if (supported) {
      resolveInfos.add(resolveInfo);
    }
    final PackageManager packageManager = mock(PackageManager.class);
    when(packageManager.queryIntentServices(any(Intent.class), anyInt())).thenReturn(resolveInfos);

    activity = mock(FacebookActivity.class);
    when(mockLoginClient.getActivity()).thenReturn(activity);
    when(activity.getPackageManager()).thenReturn(packageManager);

    mockStatic(FacebookSdk.class);
    when(FacebookSdk.getApplicationContext()).thenReturn(activity);
  }

  private void mockCustomTabRedirectActivity(final boolean hasActivity) {
    mockStatic(Validate.class);
    when(Validate.hasCustomTabRedirectActivity(nullable(Context.class), nullable(String.class)))
        .thenReturn(hasActivity);
  }

  private LoginClient.Result testCustomTabHandles(String authenticationTokenString) {
    mockCustomTabRedirectActivity(true);
    CustomTabLoginMethodHandler handler = new CustomTabLoginMethodHandler(mockLoginClient);

    final Bundle bundle = new Bundle();
    bundle.putString("access_token", ACCESS_TOKEN);
    bundle.putString(AuthenticationToken.AUTHENTICATION_TOKEN_KEY, authenticationTokenString);
    bundle.putString("expires_in", String.format("%d", EXPIRES_IN_DELTA));
    bundle.putString("code", "Something else");
    bundle.putString("signed_request", SIGNED_REQUEST_STR);
    handler.onComplete(request, bundle, null);

    final ArgumentCaptor<LoginClient.Result> resultArgumentCaptor =
        ArgumentCaptor.forClass(LoginClient.Result.class);
    verify(mockLoginClient, times(1)).completeAndValidate(resultArgumentCaptor.capture());

    final LoginClient.Result result = resultArgumentCaptor.getValue();
    assertNotNull(result);
    assertEquals(LoginClient.Result.Code.SUCCESS, result.code);

    final AccessToken token = result.token;
    assertNotNull(token);
    assertEquals(ACCESS_TOKEN, token.getToken());
    assertDateDiffersWithinDelta(new Date(), token.getExpires(), EXPIRES_IN_DELTA * 1000, 1000);
    TestUtils.assertSamePermissions(PERMISSIONS, token.getPermissions());

    return result;
  }
}

File: facebook-common/src/test/kotlin/com/facebook/login/CustomTabLoginMethodHandlerTest.kt
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 * All rights reserved.
 *
 * This source code is licensed under the license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.login

import android.app.Activity
import android.content.Intent
import android.content.pm.PackageManager
import android.content.pm.ResolveInfo
import android.content.pm.ServiceInfo
import android.os.Bundle
import androidx.fragment.app.Fragment
import com.facebook.AccessToken
import com.facebook.AccessTokenSource
import com.facebook.AuthenticationToken
import com.facebook.CustomTabMainActivity
import com.facebook.FacebookActivity
import com.facebook.FacebookException
import com.facebook.FacebookOperationCanceledException
import com.facebook.FacebookSdk
import com.facebook.internal.Validate
import com.facebook.internal.security.OidcSecurityUtil
import com.facebook.login.AuthenticationTokenTestUtil.getEncodedAuthTokenStringForTest
import org.assertj.core.api.Assertions.assertThat
import org.junit.Before
import org.junit.Test
import org.mockito.kotlin.any
import org.mockito.kotlin.anyOrNull
import org.mockito.kotlin.argumentCaptor
import org.mockito.kotlin.mock
import org.mockito.kotlin.verify
import org.mockito.kotlin.whenever
import org.powermock.api.mockito.PowerMockito
import org.powermock.core.classloader.annotations.PrepareForTest
import org.powermock.reflect.Whitebox
import java.util.Date

@PrepareForTest(Validate::class, FacebookSdk::class, OidcSecurityUtil::class)
class CustomTabLoginMethodHandlerTest : LoginHandlerTestCase() {
    private lateinit var request: LoginClient.Request
    private lateinit var mockAccessTokenCompanion: AccessToken.Companion

    @Before
    fun setUp() {
        PowerMockito.mockStatic(FacebookSdk::class.java)
        whenever(FacebookSdk.isInitialized()).thenReturn(true)
        whenever(FacebookSdk.getApplicationId()).thenReturn(AuthenticationTokenTestUtil.APP_ID)

        mockAccessTokenCompanion = mock()
        whenever(mockAccessTokenCompanion.getCurrentAccessToken()).thenReturn(null)
        Whitebox.setInternalState(AccessToken::class.java, "Companion", mockAccessTokenCompanion)

        val fragment: Fragment = mock<LoginFragment>()
        whenever(mockLoginClient.fragment).thenReturn(fragment)
        request = createRequest()

        // mock and bypass signature verification
        PowerMockito.mockStatic(OidcSecurityUtil::class.java)
        whenever(OidcSecurityUtil.getRawKeyFromEndPoint(any())).thenReturn("key")
        whenever(OidcSecurityUtil.getPublicKeyFromString(any())).thenReturn(mock())
        whenever(OidcSecurityUtil.verify(any(), any(), any())).thenReturn(true)
    }

    @Test
    fun testCustomTabHandlesSuccess() {
        testCustomTabHandles(encodedAuthTokenString)
    }

    @Test
    fun testCustomTabHandlesSuccessWithEmptyAuthenticationToken() {
        val result = testCustomTabHandles("")
        val authenticationToken = result.authenticationToken
        assertThat(authenticationToken).isNull()
    }

    @Test
    fun testCustomTabHandlesSuccessWithNoAuthenticationToken() {
        val result = testCustomTabHandles(null)
        val authenticationToken = result.authenticationToken
        assertThat(authenticationToken).isNull()
    }

    @Test
    fun testIdTokenWithNonceCustomTabHandlesSuccess() {
        mockCustomTabRedirectActivity(true)
        val requestWithNonce = createRequestWithNonce()
        val handler = CustomTabLoginMethodHandler(mockLoginClient)
        val expectedIdTokenString = getEncodedAuthTokenStringForTest()
        val bundle = Bundle()
        bundle.putString("access_token", ACCESS_TOKEN)
        bundle.putString(AuthenticationToken.AUTHENTICATION_TOKEN_KEY, expectedIdTokenString)
        bundle.putString("expires_in", String.format("%d", EXPIRES_IN_DELTA))
        bundle.putString("signed_request", SIGNED_REQUEST_STR)
        handler.onComplete(requestWithNonce, bundle, null)
        val resultArgumentCaptor = argumentCaptor<LoginClient.Result>()
        verify(mockLoginClient).completeAndValidate(resultArgumentCaptor.capture())
        val result = checkNotNull(resultArgumentCaptor.firstValue)
        assertThat(result.code).isEqualTo(LoginClient.Result.Code.SUCCESS)
        val idToken = checkNotNull(result.authenticationToken)
        assertThat(idToken.token).isEqualTo(expectedIdTokenString)
        val token = checkNotNull(result.token)
        assertThat(token.token).isEqualTo(ACCESS_TOKEN)
        assertDateDiffersWithinDelta(Date(), token.expires, EXPIRES_IN_DELTA * 1000, 1000)
        assertThat(PERMISSIONS).isEqualTo(token.permissions)
    }

    @Test
    fun testIGCustomTabHandlesSuccess() {
        mockCustomTabRedirectActivity(true)
        val igRequest = createIGWebRequest()
        val handler = CustomTabLoginMethodHandler(mockLoginClient)
        val bundle = Bundle()
        bundle.putString("access_token", ACCESS_TOKEN)
        bundle.putString("graph_domain", "instagram")
        bundle.putString("signed_request", SIGNED_REQUEST_STR)
        handler.onComplete(igRequest, bundle, null)
        val resultArgumentCaptor = argumentCaptor<LoginClient.Result>()
        verify(mockLoginClient).completeAndValidate(resultArgumentCaptor.capture())
        val result = resultArgumentCaptor.firstValue
        assertThat(result).isNotNull
        assertThat(result.code).isEqualTo(LoginClient.Result.Code.SUCCESS)
        val token = checkNotNull(result.token)
        assertThat(token.token).isEqualTo(ACCESS_TOKEN)
        assertThat(token.userId).isEqualTo(USER_ID)
        assertThat(token.graphDomain).isEqualTo("instagram")
        assertThat(token.source).isEqualTo(AccessTokenSource.INSTAGRAM_CUSTOM_CHROME_TAB)
        assertThat(PERMISSIONS).isEqualTo(token.permissions)
    }

    @Test
    fun testCustomTabHandlesCancel() {
        mockCustomTabRedirectActivity(true)
        val handler = CustomTabLoginMethodHandler(mockLoginClient)
        handler.onComplete(request, null, FacebookOperationCanceledException())
        val resultArgumentCaptor = argumentCaptor<LoginClient.Result>()
        verify(mockLoginClient).completeAndValidate(resultArgumentCaptor.capture())
        val result = checkNotNull(resultArgumentCaptor.firstValue)
        assertThat(result.code).isEqualTo(LoginClient.Result.Code.CANCEL)
        assertThat(result.token).isNull()
        assertThat(result.errorMessage).isNotNull
    }

    @Test
    fun testCustomTabHandlesError() {
        mockCustomTabRedirectActivity(true)
        val handler = CustomTabLoginMethodHandler(mockLoginClient)
        handler.onComplete(request, null, FacebookException(ERROR_MESSAGE))
        val resultArgumentCaptor = argumentCaptor<LoginClient.Result>()
        verify(mockLoginClient).completeAndValidate(resultArgumentCaptor.capture())
        val result = checkNotNull(resultArgumentCaptor.firstValue)
        assertThat(result.code).isEqualTo(LoginClient.Result.Code.ERROR)
        assertThat(result.token).isNull()
        assertThat(result.errorMessage).isNotNull
        assertThat(result.errorMessage).isEqualTo(ERROR_MESSAGE)
    }

    @Test
    fun testTryAuthorizeNeedsRedirectActivity() {
        mockChromeCustomTabsSupported(true, CHROME_PACKAGE)
        mockCustomTabRedirectActivity(true)
        val handler = CustomTabLoginMethodHandler(mockLoginClient)
        assertThat(handler.tryAuthorize(request)).isEqualTo(1)
    }

    @Test
    fun testTryAuthorizeWithChromePackage() {
        mockCustomTabRedirectActivity(true)
        mockChromeCustomTabsSupported(true, CHROME_PACKAGE)
        val handler = CustomTabLoginMethodHandler(mockLoginClient)
        assertThat(handler.tryAuthorize(request)).isEqualTo(1)
    }

    @Test
    fun testTryAuthorizeWithChromeBetaPackage() {
        mockCustomTabRedirectActivity(true)
        mockChromeCustomTabsSupported(true, BETA_PACKAGE)
        val handler = CustomTabLoginMethodHandler(mockLoginClient)
        assertThat(handler.tryAuthorize(request)).isEqualTo(1)
    }

    @Test
    fun testTryAuthorizeWithChromeDevPackage() {
        mockCustomTabRedirectActivity(true)
        mockChromeCustomTabsSupported(true, DEV_PACKAGE)
        val handler = CustomTabLoginMethodHandler(mockLoginClient)
        assertThat(handler.tryAuthorize(request)).isEqualTo(1)
    }

    private fun mockChromeCustomTabsSupported(supported: Boolean, packageName: String) {
        val resolveInfos: MutableList<ResolveInfo> = ArrayList()
        val resolveInfo = ResolveInfo()
        val serviceInfo = ServiceInfo()
        serviceInfo.packageName = packageName
        resolveInfo.serviceInfo = serviceInfo
        if (supported) {
            resolveInfos.add(resolveInfo)
        }
        val packageManager = mock<PackageManager>()
        whenever(packageManager.queryIntentServices(any<Intent>(), any<Int>())).thenReturn(
            resolveInfos
        )
        activity = mock<FacebookActivity>()
        whenever(mockLoginClient.activity).thenReturn(activity)
        whenever(activity.packageManager).thenReturn(packageManager)
        PowerMockito.mockStatic(FacebookSdk::class.java)
        whenever(FacebookSdk.getApplicationContext()).thenReturn(activity)
    }

    private fun mockCustomTabRedirectActivity(hasActivity: Boolean) {
        PowerMockito.mockStatic(Validate::class.java)
        whenever(Validate.hasCustomTabRedirectActivity(anyOrNull(), anyOrNull()))
            .thenReturn(hasActivity)
    }

    private fun testCustomTabHandles(authenticationTokenString: String?): LoginClient.Result {
        val requestWithNonce = createRequestWithNonce()
        mockCustomTabRedirectActivity(true)
        val handler = CustomTabLoginMethodHandler(mockLoginClient)
        val bundle = Bundle()
        bundle.putString("access_token", ACCESS_TOKEN)
        bundle.putString(AuthenticationToken.AUTHENTICATION_TOKEN_KEY, authenticationTokenString)
        bundle.putString("expires_in", String.format("%d", EXPIRES_IN_DELTA))
        bundle.putString("code", "Something else")
        bundle.putString("signed_request", SIGNED_REQUEST_STR)
        handler.onComplete(requestWithNonce, bundle, null)
        val resultArgumentCaptor = argumentCaptor<LoginClient.Result>()
        verify(mockLoginClient).completeAndValidate(resultArgumentCaptor.capture())
        val result = checkNotNull(resultArgumentCaptor.firstValue)
        assertThat(result.code).isEqualTo(LoginClient.Result.Code.SUCCESS)
        val token = checkNotNull(result.token)
        assertThat(token.token).isEqualTo(ACCESS_TOKEN)
        assertDateDiffersWithinDelta(Date(), token.expires, EXPIRES_IN_DELTA * 1000, 1000)
        assertThat(PERMISSIONS).isEqualTo(token.permissions)
        return result
    }

    @Test
    fun `test receiving no browser exception as activity result`() {
        mockCustomTabRedirectActivity(true)
        val data = Intent()
        data.putExtra(CustomTabMainActivity.NO_ACTIVITY_EXCEPTION, true)
        val handler = CustomTabLoginMethodHandler(mockLoginClient)
        assertThat(handler.onActivityResult(1, Activity.RESULT_CANCELED, data)).isFalse
    }

    @Test
    fun `test receiving user cancel result`() {
        mockCustomTabRedirectActivity(true)
        whenever(mockLoginClient.pendingRequest).thenReturn(request)
        val handler = CustomTabLoginMethodHandler(mockLoginClient)
        assertThat(handler.onActivityResult(1, Activity.RESULT_CANCELED, null)).isFalse
        val resultCaptor = argumentCaptor<LoginClient.Result>()
        verify(mockLoginClient).completeAndValidate(resultCaptor.capture())
        val result = resultCaptor.firstValue
        assertThat(result.code).isEqualTo(LoginClient.Result.Code.CANCEL)
        assertThat(result.errorMessage).isEqualTo(LoginMethodHandler.USER_CANCELED_LOG_IN_ERROR_MESSAGE)
    }

    @Test
    fun `test receiving user ok result`() {
        mockCustomTabRedirectActivity(true)
        whenever(mockLoginClient.pendingRequest).thenReturn(request)
        val handler = CustomTabLoginMethodHandler(mockLoginClient)
        assertThat(handler.onActivityResult(1, Activity.RESULT_OK, null)).isTrue
    }

    @Test
    fun testAddExtraParametersWithRedirectURI() {
        mockCustomTabRedirectActivity(true)
        val handler = CustomTabLoginMethodHandler(mockLoginClient)
        val testRedirectURI = "https://example.com/redirect"

        // Create a request with redirect URI
        val requestWithRedirectURI = createRequestWithRedirectURI(testRedirectURI)

        // Create initial parameters bundle
        val initialParameters = Bundle()
        initialParameters.putString("existing_param", "existing_value")

        // Call addExtraParameters using reflection to access protected method
        val updatedParameters = Whitebox.invokeMethod<Bundle>(
            handler,
            "addExtraParameters",
            initialParameters,
            requestWithRedirectURI
        )

        // Verify that https_redirect_uri parameter is included
        assertThat(updatedParameters.getString("https_redirect_uri")).isEqualTo(testRedirectURI)
        // Verify existing parameters are preserved
        assertThat(updatedParameters.getString("existing_param")).isEqualTo("existing_value")
    }

    @Test
    fun testAddExtraParametersWithoutRedirectURI() {
        mockCustomTabRedirectActivity(true)
        val handler = CustomTabLoginMethodHandler(mockLoginClient)

        // Create a request without redirect URI (passing null as the redirectURI parameter)
        val requestWithoutRedirectURI = createRequestWithRedirectURI(null)

        // Create initial parameters bundle
        val initialParameters = Bundle()
        initialParameters.putString("existing_param", "existing_value")

        // Call addExtraParameters using reflection to access protected method
        val updatedParameters = Whitebox.invokeMethod<Bundle>(
            handler,
            "addExtraParameters",
            initialParameters,
            requestWithoutRedirectURI
        )

        // Verify that https_redirect_uri parameter is NOT included
        assertThat(updatedParameters.containsKey("https_redirect_uri")).isFalse()
        // Verify existing parameters are preserved
        assertThat(updatedParameters.getString("existing_param")).isEqualTo("existing_value")
    }

    @Test
    fun testAddExtraParametersWithEmptyRedirectURI() {
        mockCustomTabRedirectActivity(true)
        val handler = CustomTabLoginMethodHandler(mockLoginClient)

        // Create a request with empty redirect URI
        val requestWithEmptyRedirectURI = createRequestWithRedirectURI("")

        // Create initial parameters bundle
        val initialParameters = Bundle()
        initialParameters.putString("existing_param", "existing_value")

        // Call addExtraParameters using reflection to access protected method
        val updatedParameters = Whitebox.invokeMethod<Bundle>(
            handler,
            "addExtraParameters",
            initialParameters,
            requestWithEmptyRedirectURI
        )

        // Verify that https_redirect_uri parameter is NOT included
        assertThat(updatedParameters.containsKey("https_redirect_uri")).isFalse()
        // Verify existing parameters are preserved
        assertThat(updatedParameters.getString("existing_param")).isEqualTo("existing_value")
    }

    @Test
    fun testTryAuthorizeWithRedirectURI() {
        mockCustomTabRedirectActivity(true)
        mockChromeCustomTabsSupported(true, CHROME_PACKAGE)
        val handler = CustomTabLoginMethodHandler(mockLoginClient)
        val testRedirectURI = "https://example.com/custom/redirect"

        // Create a request with redirect URI
        val requestWithRedirectURI = createRequestWithRedirectURI(testRedirectURI)

        // Call tryAuthorize - this should process the redirect URI through addExtraParameters
        val result = handler.tryAuthorize(requestWithRedirectURI)

        // Verify that the operation completed successfully (should return 1 for success)
        assertThat(result).isEqualTo(1)
        // Verify that the request's redirect URI is preserved
        assertThat(requestWithRedirectURI.redirectURI).isEqualTo(testRedirectURI)
    }

    private fun createRequestWithRedirectURI(redirectURI: String?): LoginClient.Request {
        return LoginClient.Request(
            LoginBehavior.NATIVE_WITH_FALLBACK,
            HashSet(PERMISSIONS),
            DefaultAudience.FRIENDS,
            "rerequest",
            "1234",
            "5678",
            LoginTargetApp.FACEBOOK,
            AuthenticationTokenTestUtil.NONCE,
            CODE_VERIFIER,
            CODE_CHALLENGE,
            CodeChallengeMethod.S256,
            redirectURI)
    }

    companion object {
        private const val SIGNED_REQUEST_STR =
            "ggarbage.eyJhbGdvcml0aG0iOiJITUFDSEEyNTYiLCJjb2RlIjoid2h5bm90IiwiaXNzdWVkX2F0IjoxNDIyNTAyMDkyLCJ1c2VyX2lkIjoiMTIzIn0"
        private const val CHROME_PACKAGE = "com.android.chrome"
        private const val DEV_PACKAGE = "com.chrome.dev"
        private const val BETA_PACKAGE = "com.chrome.beta"
    }
}
